airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/CatalogHelpers.java

```diff
@@ -124,31 +134,151 @@ public static Set<String> getTopLevelFieldNames(final ConfiguredAirbyteStream st
   }
 
   /**
-   * @param node any json node
+   * @param jsonSchema - a JSONSchema node
    * @return a set of all keys for all objects within the node
    */
   @VisibleForTesting
-  protected static Set<String> getAllFieldNames(final JsonNode node) {
-    final Set<String> allFieldNames = new HashSet<>();
-
-    if (node.has("properties")) {
-      final JsonNode properties = node.get("properties");
-      final Iterator<String> fieldNames = properties.fieldNames();
-      while (fieldNames.hasNext()) {
-        final String fieldName = fieldNames.next();
-        allFieldNames.add(fieldName);
-        final JsonNode fieldValue = properties.get(fieldName);
-        if (fieldValue.isObject()) {
-          allFieldNames.addAll(getAllFieldNames(fieldValue));
-        }
+  protected static Set<String> getAllFieldNames(final JsonNode jsonSchema) {
+    return getFullyQualifiedFieldNamesWithTypes(jsonSchema)
+        .stream()
+        .map(Pair::getLeft)
+        // only need field name, not fully qualified name
+        .map(MoreLists::last)
+        .flatMap(Optional::stream)
+        .collect(Collectors.toSet());
+  }
+
+  /**
+   * Extracts all fields and their schemas from a JSONSchema. This method returns values in
+   * depth-first search preorder. It short circuits at oneOfs--in other words, child fields of a oneOf
+   * are not returned.
+   *
+   * @param jsonSchema - a JSONSchema node
+   * @return a list of all keys for all objects within the node. ordered in depth-first search
+   *         preorder.
+   */
+  @VisibleForTesting
+  protected static List<Pair<List<String>, JsonNode>> getFullyQualifiedFieldNamesWithTypes(final JsonNode jsonSchema) {
+    // if this were ever a performance issue, it could be replaced with a trie. this seems unlikely
+    // however.
+    final Set<List<String>> fieldNamesThatAreOneOfs = new HashSet<>();
+
+    return JsonSchemas.traverseJsonSchemaWithCollector(jsonSchema, (node, basicPath) -> {
+      final List<String> fieldName = basicPath.stream().filter(fieldOrList -> !fieldOrList.isList()).map(FieldNameOrList::getFieldName).toList();
+      return Pair.of(fieldName, node);
+    })
+        .stream()
+        // first node is the original object.
+        .skip(1)
+        .filter(fieldWithSchema -> filterChildrenOfFoneOneOf(fieldWithSchema.getLeft(), fieldWithSchema.getRight(), fieldNamesThatAreOneOfs))
```

Comment:
I don't think this is an issue here but I think we should avoid using mutable variable in a lambda because of concurrency issues (for example with a parallel stream, it might fail).